pkg/admission/customresourcedefinition/admission.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package customresourcedefinition
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/kcp-dev/apimachinery/pkg/logicalcluster"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+
+	"github.com/kcp-dev/kcp/pkg/reconciler/apis/apibinding"
+)
+
+const (
+	PluginName        = "tenancy.kcp.dev/BoundCustomResourceDefinition"
+	boundCRDWorkspace = "apis.kcp.dev/bound-crd"
+)
+
+func Register(plugins *admission.Plugins) {
+	plugins.Register(PluginName,
+		func(_ io.Reader) (admission.Interface, error) {
+			return &customResourceDefinitionBound{
+				Handler: admission.NewHandler(admission.Create, admission.Update),
+			}, nil
+		})
+}
+
+type customResourceDefinitionBound struct {
+	*admission.Handler
+}
+
+// Ensure that the required admission interfaces are implemented.
+var _ = admission.ValidationInterface(&customResourceDefinitionBound{})
+
+// Validate ensures that
+// - has a valid type
+// - has valid initializers when transitioning to initializing
+func (o *customResourceDefinitionBound) Validate(ctx context.Context, a admission.Attributes, _ admission.ObjectInterfaces) (err error) {
+	if a.GetResource().GroupResource() != apiextensionsv1.Resource("customresourcedefinitions") {
+		return nil
+	}
+
+	if a.GetKind() != apiextensionsv1.Kind("CustomResourceDefinition").WithVersion("v1") {
+		klog.Infoln("Here at validation did not pass GVK checks", a.GetKind(), apiextensionsv1.Kind("CustomResourceDefinition").WithVersion("v1"))
+		return nil
+	}
+	cw, ok := a.GetObject().(*apiextensions.CustomResourceDefinition)
+	if !ok {
+		klog.Infoln(fmt.Sprintf("unexpected type %T", a.GetObject()))
+		return fmt.Errorf("unexpected type %T", a.GetObject())
+	}
+	klog.Infoln("Here at validation passed conversion to CRD")
+
+	if _, ok := cw.GetAnnotations()[boundCRDWorkspace]; !ok {
+		// if no annotation pass through
+		return nil
+	}
+
+	name, err := request.ClusterNameFrom(ctx)
```

Comment:
I wonder if it would make more sense to get the cluster name from the CRD? i.e. `logicalcluster.From(cw)`? @sttts WDYT?

Reply:
I was worried that it would not be sent, but if it will then I could change.

Reply:
Oh right, this is not always objects coming out of etcd. Nevermind!